// @threlte/core/dist/index.js
export const VERSION = 8;
// canvas component
export { default as Canvas } from './Canvas.svelte';
// components (v6)
export { T, extend } from './components/T/T.js';
// plugins
export { injectPlugin } from './plugins/injectPlugin.js';
// hooks
export { useThrelte } from './context/compounds/useThrelte.js';
export { useStage } from './hooks/useStage.js';
export { useTask } from './hooks/useTask.svelte.js';
export { useThrelteUserContext } from './hooks/useThrelteUserContext.js';
// useLoader
export { useLoader } from './hooks/useLoader.js';
export { createThrelteContext } from './context/createThrelteContext.svelte.js';
export { createCacheContext, useCache } from './context/fragments/cache.js';
export { createCameraContext, useCamera } from './context/fragments/camera.js';
export { createDOMContext, useDOM } from './context/fragments/dom.js';
export { createDisposalContext, useDisposal } from './context/fragments/disposal.js';
export { createParentContext, useParent } from './context/fragments/parent.js';
export { createParentObject3DContext, useParentObject3D } from './context/fragments/parentObject3D.js';
export { createRendererContext, useRenderer } from './context/fragments/renderer.svelte.js';
export { createSceneContext, useScene } from './context/fragments/scene.js';
export { createSchedulerContext, useScheduler } from './context/fragments/scheduler.svelte.js';
export { createUserContext } from './context/fragments/user.js';
// utils
export * from './utilities/index.js';
