// @ascorbic/bluesky-loader/dist/index.js
// src/index.ts
import { AtpAgent } from "@atproto/api";

// src/utils.ts
import {
  RichText
} from "@atproto/api";
var escapeMap = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
  "'": "&#39;"
};
var escapeHTML = (str) => str?.replace(/[&<>"']/g, (match) => escapeMap[match] || match) ?? "";
var nl2br = (str) => str?.replaceAll("\n", "<br/ >\n") ?? "";
function renderPostAsHtml(post, transformNewlines = true) {
  const rt = new RichText(post.record);
  let html = "";
  for (const segment of rt.segments()) {
    if (segment.isLink()) {
      html += `<a href="${escapeHTML(segment.link?.uri)}">${escapeHTML(segment.text)}</a>`;
    } else if (segment.isMention()) {
      html += `<a href="https://bsky.app/profile/${escapeHTML(segment.mention.did)}">${escapeHTML(segment.text)}</a>`;
    } else if (segment.isTag()) {
      html += `<a href="https://bsky.app/hashtag/${escapeHTML(segment.tag.tag)}">#${escapeHTML(segment.tag.tag)}</a>`;
    } else {
      html += transformNewlines ? nl2br(escapeHTML(segment.text)) : escapeHTML(segment.text);
    }
  }
  return html;
}

// src/schema.ts
import { z } from "astro/zod";
var EmbedExternalSchema = z.object({
  $type: z.literal("app.bsky.embed.external"),
  external: z.object({
    uri: z.string().url(),
    title: z.string().optional(),
    description: z.string().optional(),
    thumb: z.object({
      $type: z.literal("blob"),
      ref: z.object({ $link: z.string() }),
      mimeType: z.string(),
      size: z.number()
    }).optional()
  })
});
var EmbedExternalViewSchema = z.object({
  $type: z.literal("app.bsky.embed.external#view"),
  external: z.object({
    uri: z.string().url(),
    title: z.string().optional(),
    description: z.string().optional(),
    thumb: z.string().url().optional()
    // Direct URL for view
  })
});
var EmbedRecordSchema = z.object({
  $type: z.literal("app.bsky.embed.record"),
  record: z.object({
    uri: z.string(),
    cid: z.string()
  })
});
var EmbedImagesSchema = z.object({
  $type: z.literal("app.bsky.embed.images"),
  images: z.array(
    z.object({
      image: z.object({
        $type: z.literal("blob"),
        ref: z.object({ $link: z.string() }),
        mimeType: z.string(),
        size: z.number()
      }),
      alt: z.string()
    })
  )
});
var EmbedImagesViewSchema = z.object({
  $type: z.literal("app.bsky.embed.images#view"),
  images: z.array(
    z.object({
      thumb: z.string().url(),
      fullsize: z.string().url(),
      alt: z.string().optional()
    })
  )
});
var EmbedListSchema = z.object({
  $type: z.literal("app.bsky.embed.list"),
  list: z.object({
    title: z.string(),
    items: z.array(
      z.object({
        uri: z.string(),
        label: z.string()
      })
    )
  })
});
var EmbedListViewSchema = z.object({
  $type: z.literal("app.bsky.embed.list#view"),
  list: z.object({
    title: z.string(),
    items: z.array(
      z.object({
        uri: z.string(),
        label: z.string()
      })
    )
  })
});
var EmbedStarterPackSchema = z.object({
  $type: z.literal("app.bsky.embed.starterPack"),
  starterPack: z.object({
    title: z.string(),
    description: z.string().optional(),
    members: z.array(
      z.object({
        did: z.string(),
        handle: z.string(),
        displayName: z.string().optional(),
        avatar: z.string().url().optional()
      })
    )
  })
});
var EmbedRecordViewSchema = z.object({
  $type: z.literal("app.bsky.embed.record#view"),
  record: z.object({
    uri: z.string(),
    cid: z.string(),
    author: z.object({
      did: z.string(),
      handle: z.string(),
      displayName: z.string(),
      avatar: z.string().url().optional()
    }),
    value: z.object({
      $type: z.string(),
      createdAt: z.string().datetime(),
      text: z.string().optional()
    })
  })
});
var EmbedRecordWithMediaSchema = z.object({
  $type: z.literal("app.bsky.embed.recordWithMedia"),
  record: z.object({
    uri: z.string(),
    cid: z.string(),
    author: z.object({
      did: z.string(),
      handle: z.string(),
      displayName: z.string(),
      avatar: z.string().url().optional()
    }),
    value: z.object({
      $type: z.string(),
      createdAt: z.string().datetime(),
      text: z.string().optional()
    })
  }),
  media: z.union([
    EmbedExternalSchema,
    // External embed type
    EmbedImagesSchema
    // Image embed type
  ])
});
var EmbedRecordWithMediaViewSchema = z.object({
  $type: z.literal("app.bsky.embed.recordWithMedia#view"),
  record: z.object({
    uri: z.string(),
    cid: z.string(),
    author: z.object({
      did: z.string(),
      handle: z.string(),
      displayName: z.string(),
      avatar: z.string().url().optional()
    }),
    value: z.object({
      $type: z.string(),
      createdAt: z.string().datetime(),
      text: z.string().optional()
    })
  }),
  media: z.union([
    EmbedExternalViewSchema,
    // External embed view type
    EmbedImagesViewSchema
    // Image embed view type
  ])
});
var EmbedStarterPackViewSchema = z.object({
  $type: z.literal("app.bsky.embed.starterPack#view"),
  starterPack: z.object({
    title: z.string(),
    description: z.string().optional(),
    members: z.array(
      z.object({
        did: z.string(),
        handle: z.string(),
        displayName: z.string().optional(),
        avatar: z.string().url().optional()
      })
    )
  })
});
var UnknownEmbedSchema = z.object({
  $type: z.string()
}).passthrough();
var EmbedSchema = z.union([
  EmbedExternalSchema,
  EmbedRecordSchema,
  EmbedRecordWithMediaSchema,
  EmbedImagesSchema,
  EmbedListSchema,
  EmbedStarterPackSchema,
  EmbedExternalViewSchema,
  EmbedRecordViewSchema,
  EmbedRecordWithMediaViewSchema,
  EmbedImagesViewSchema,
  EmbedListViewSchema,
  EmbedStarterPackViewSchema,
  UnknownEmbedSchema
  // Catch-all for unknown embed types
]);
var PostSchema = z.object({
  uri: z.string(),
  cid: z.string(),
  author: z.object({
    did: z.string(),
    handle: z.string(),
    displayName: z.string(),
    avatar: z.string().url().optional(),
    associated: z.object({
      chat: z.object({
        allowIncoming: z.enum(["following", "all", "none"]).optional()
      }).optional()
    }).optional(),
    labels: z.array(z.any()).optional(),
    createdAt: z.string().datetime()
  }),
  record: z.object({
    $type: z.literal("app.bsky.feed.post"),
    createdAt: z.string().datetime(),
    langs: z.array(z.string()).optional(),
    text: z.string().optional(),
    reply: z.object({
      parent: z.object({
        cid: z.string(),
        uri: z.string()
      }),
      root: z.object({
        cid: z.string(),
        uri: z.string()
      })
    }).optional(),
    embed: EmbedSchema.optional(),
    facets: z.array(
      z.object({
        index: z.object({
          byteStart: z.number().nonnegative(),
          byteEnd: z.number().nonnegative()
        }),
        features: z.array(
          z.union([
            z.object({
              $type: z.literal("app.bsky.richtext.facet#mention"),
              did: z.string()
            }),
            z.object({
              $type: z.literal("app.bsky.richtext.facet#link"),
              uri: z.string().url()
            }),
            z.object({
              $type: z.literal("app.bsky.richtext.facet#tag"),
              tag: z.string()
            }),
            z.object({}).passthrough()
            // Fallback for unknown facet types
          ])
        )
      })
    ).optional()
  }).passthrough(),
  // Allow unknown properties
  embed: EmbedSchema.optional(),
  replyCount: z.number().nonnegative(),
  repostCount: z.number().nonnegative(),
  likeCount: z.number().nonnegative(),
  quoteCount: z.number().nonnegative(),
  indexedAt: z.string().datetime(),
  labels: z.array(z.any()).optional()
});

// src/index.ts
var authorFeedLoader = ({
  identifier,
  filter,
  limit
}) => {
  return {
    name: "bluesky-loader",
    schema: PostSchema,
    async load({ store, logger, meta, parseData }) {
      const agent = new AtpAgent({ service: "https://public.api.bsky.app" });
      try {
        const mostRecent = meta.get("lastFetched") || 0;
        let cursor = void 0;
        let first;
        let count = 0;
        fetching: do {
          const { data } = await agent.getAuthorFeed({
            actor: identifier,
            filter,
            cursor,
            limit: 100
          });
          for (const { post } of data.feed) {
            if (mostRecent && mostRecent === post.cid || limit && count >= limit) {
              count++;
              break fetching;
            }
            count++;
            if (!first) {
              first = post.cid;
            }
            store.set({
              id: post.uri,
              data: await parseData({
                id: post.uri,
                // Convert the post object to a plain object
                data: JSON.parse(JSON.stringify(post))
              }),
              rendered: {
                html: renderPostAsHtml(post)
              }
            });
          }
          cursor = data.cursor;
          logger.info(`Fetched ${count} posts`);
        } while (cursor);
        if (first) {
          meta.set("lastFetched", first);
        }
      } catch (error) {
        logger.error(
          `Failed to load Bluesky posts. ${error.message}`
        );
      }
    }
  };
};
export {
  authorFeedLoader,
  renderPostAsHtml
};
